pennylane/qaoa/layers.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Methods that define cost and mixer layers for use in QAOA workflows.
+"""
+import pennylane as qml
+from pennylane.operation import Tensor
+
+
+def _diagonal_terms(hamiltonian):
+    r"""Checks if all terms in a Hamiltonian are products of diagonal Pauli gates
+    (:class:`~.PauliZ` and :class:`~.Identity`).
+
+    Args:
+        hamiltonian (.Hamiltonian): The Hamiltonian being checked
+
+    Returns:
+        bool: ``True`` if all terms are products of diagonal Pauli gates, ``False`` otherwise
+    """
+    val = True
+
+    for i in hamiltonian.ops:
+        i = Tensor(i) if not isinstance(i, Tensor) else i
+        for j in i.obs:
+            if j.name not in ("PauliZ", "Identity"):
+                val = False
+                break
+
+    return val
+
+
+def cost_layer(gamma, hamiltonian):
```

Comment:
I know it is already an abuse of style, but the other templates use the "CapitalizedWords" class-like case style (even though they are also functions). Is there a motivation for breaking this tradition here?

Reply:
This is a good point! @ixfoduap @josh146 what do you think?

Reply:
@trbromley I talked to Josh, and he is Ok with either or. @ixfoduap do you have a preference?

Reply:
Thanks for the suggestion Tom! I view this more as a function inside qaoa than a template, so I prefer not using capitalized naming here